### ISSUE TITLE
File system for *store.db.* kept on growing after loading and deleting nodes, relations and property from same csv file again and again

### ISSUE THREAD

**Issue by 64BitAsura** [neo4j/neo4j#12035]
## Guidelines
- Neo4j version: 3.4.1
- Operating system: Docker image **neo4j:3.4.1**
- API/Driver: (Noe)
- **Steps to reproduce**
  Loading node and relation from CSV file about million of them and deleting them completely, repeat these step several times,

- Expected behavior
On insert of million node and relation
   /data/graph.db/neostore.nodestore.db ( size almost intact unless too much data is loaded )
  /data/graph.db/neostore.nodestore.db.id ( size need to reduce since because of recycling of node id)
  same goes for neostore.relationshipstore.db and neostore.relationshipstore.db.id respectively 
- Actual behavior
   /data/graph.db/neostore.nodestore.db ( size is increasing for same load after deletion )
  /data/graph.db/neostore.nodestore.db.id ( size is not reducing at all)
  same goes for neostore.relationshipstore.db and neostore.relationshipstore.db.id respectively 

Additionally, include (as appropriate) log-files, stacktraces, and other debug output.
But /data/graph.db/neostore.propertystore.db and /data/graph.db/neostore.propertystore.db.id working as expected in runtime

**Comment by geirgp** [comment]
Could this be a side-effect of the transaction log pointed out here? https://stackoverflow.com/questions/46771529/neo4j-store-size-increases-after-major-delete-operation

**Comment by tinwelint** [comment]
I take it that the transaction logs size isn't included since you mention the `*store.db*` pattern.

tl;dr this is a known issue which will be fixed in 4.0

Up to and including 3.5 record reuse is circumstantial and most record stores will not reuse ids until after a clean restart - this is why you see an ever growing database. Also if the database isn't cleanly shutdown in the restart the .id files are deleted, so it's a bit sensitive.

However in 4.0 this has been completely reworked so all deleted records, regardless of type can be reused almost immediately after being deleted without restart, and it doesn't matter whether or not the database is cleanly shut down or even crashes, the records will still be reused.
